### ISSUE TITLE
[BUG] Invalid string type when reading declarative `setup.cfg` under Python 2

### ISSUE THREAD

**Issue by webknjaz** [pypa/setuptools#1136]
Hi,

I'm seeing the following error:
```python
error: 'egg_base' must be a directory name (got `src`)
```

The config contains:
```ini
[options]
...
package_dir =
     = src
```

Contents of `package_dir` is
```python
{u'': u'src'}
```
(I've worked around this by manually reading config and patching this applying `str()` to keys and vars)

Python: 2.7
OS: GNU/Linux

The issue appears to be the same as for:
* ref #190
* ref #597

**Comment by jensens** [comment]
I can confirm this problem. 
I tried to empty the ``setup.py`` and use the declarative way with information in ``setup.cfg`` only. 
This problem is the only stopper for me to use this way for Python 2. And since our packages shall support both, it is a stopper for Python 3 as well.

If ``setup.cfg`` contains:
```
package_dir = 
    = src
```

Under Python 2 the arguments of the ``parsedict`` (from ``setuptools.config``) are unicode strings:
```
>>> type(setuptools.config.read_configuration('setup.cfg')['options']['package_dir'][''])
<type 'unicode'>
```

While under Python 3 
```
>>> type(setuptools.config.read_configuration('setup.cfg')['options']['package_dir'][''])
<class 'str'>
```

Which overall *should* be fine, but it is not. Under Python 2 a string ``<str>`` and not unicode seems to be expected. 
I did not dig this down to distutils, but when modifing line 225 in ``setuptools.config`` to turn the type into a str with ``result[key.strip()] = str(val.strip())`` all works fine.

**Comment by jaraco** [comment]
Rather than patching only `setuptools.config`, I'd like to see a patch for `setuptools` that allows for unicode to be used in `setup.py` as well. I thought the fix for #597 did that, but apparently not for all cases.

**Comment by benoit-pierre** [comment]
I believe #1062 is related. The problem with switching to Unicode for the results of reading the configuration is it will trips things like `ensure_string` or `ensure_string_list` in distutils when running under Python 2:
https://travis-ci.org/benoit-pierre/setuptools/jobs/292662535#L763

**Comment by eromoe** [comment]
I am using `win_unicode_console` to handle unicode print in cmd, seems relate,  
got same error:
```
    d:\python27\lib\site-packages\win_unicode_console\__init__.py:31: RuntimeWarning: sys.stdin.enco
ding == 'utf-8', whereas sys.stdout.encoding == None, readline hook consumer may assume they are the
 same
      readline_hook.enable(use_pyreadline=use_pyreadline)
    error: 'egg_base' must be a directory name (got `pip-egg-info`)
```

**Comment by Midnighter** [comment]
I'm seeing this issue right now with Python 2.7.12 and

```
pip        10.0.1
setuptools 39.1.0
wheel      0.31.0
```

what's the status of fixing this?

**Comment by jaraco** [comment]
Work in #1180 released as [40.7.0](https://setuptools.readthedocs.io/en/latest/history.html#v40-7-0).

**Comment by aarchiba** [comment]
We have users still reporting this with setuptools 41.4
